### PR TITLE
chore: permissions refactoring

### PR DESCRIPTION
**Pull request by tolusha**
Signed-off-by: Anatolii Bazko <abazko@redhat.com>

### What does this PR do?
Refactor permissions reconciles 

### Screenshot/screencast of this PR

N/A

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/19284

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Custom resource definition file is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#updating-custom-resource-definition-file)
- [ ] [OLM bundles are up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-olm-bundle)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: controllers/che/checluster_controller.go [hunk 1]
```diff
@@ -99,6 +99,10 @@ func NewReconciler(
 	reconcileManager.RegisterReconciler(openshiftoauth.NewOpenShiftOAuth(openShiftOAuthUser))
 	reconcileManager.RegisterReconciler(tls.NewCertificatesReconciler())
 	reconcileManager.RegisterReconciler(tls.NewTlsSecretReconciler())
+	reconcileManager.RegisterReconciler(devworkspace.NewDevWorkspaceReconciler())
+	reconcileManager.RegisterReconciler(rbac.NewRbacReconciler())
```
Comment (mmorhun) [inline controllers/che/checluster_controller.go:103]: Aren't `GatewayPermissionsReconciler` and `WorkspacePermissionsReconciler` using RBAC? We should be more clear in names.

File: pkg/deploy/dev-workspace/dev_workspace_utils.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	for _, csv := range csvList.Items {
+		if strings.HasPrefix(csv.Name, DevWorkspaceCSVNamePrefix) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func doesWebTerminalSubscriptionExist(deployContext *deploy.DeployContext) (bool, error) {
```
Comment (mmorhun) [inline pkg/deploy/dev-workspace/dev_workspace_utils.go:59]: `does` -> `is`

File: pkg/deploy/dev-workspace/dev_workspace_utils.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	// If subscriptions resource doesn't exist in cluster WTO as well will not be present
+	if !util.HasK8SResourceObject(deployContext.ClusterAPI.DiscoveryClient, SubscriptionResourceName) {
+		return false, nil
+	}
+
+	subscription := &operatorsv1alpha1.Subscription{}
+	if err := deployContext.ClusterAPI.NonCachingClient.Get(
```
Comment (mmorhun) [inline pkg/deploy/dev-workspace/dev_workspace_utils.go:66]: Looks like this invocation using `NonCachingClient` is done on each reconcile which is expensive.

File: pkg/deploy/rbac/gateway_permissions.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (gp *GatewayPermissionsReconciler) deleteGatewayPermissions(deployContext *deploy.DeployContext) (bool, error) {
+	name := gp.gatewayPermissionsName(deployContext.CheCluster)
+	if done, err := deploy.Delete(deployContext, types.NamespacedName{Name: name}, &rbacv1.ClusterRole{}); !done {
```
Comment (mmorhun) [inline pkg/deploy/rbac/gateway_permissions.go:71]: Shouldn't we delete `RoleBinding` first?

File: pkg/deploy/rbac/rbac.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+func NewRbacReconciler() *RbacReconciler {
+	return &RbacReconciler{}
+}
+
+func (c *RbacReconciler) Reconcile(ctx *deploy.DeployContext) (reconcile.Result, bool, error) {
+	// Create service account "che" for che-server component.
+	// "che" is the one which token is used to create workspace objects.
+	// Notice: Also we have on more "che-workspace" SA used by plugins like exec, terminal, metrics with limited privileges.
```
Comment (mmorhun) [inline pkg/deploy/rbac/rbac.go:34]: `on more` -> `one more`

File: pkg/deploy/rbac/rbac.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	done, err := deploy.SyncServiceAccountToCluster(ctx, deploy.CheServiceAccountName)
+	if !done {
+		return reconcile.Result{Requeue: true}, false, err
+	}
+
+	if len(ctx.CheCluster.Spec.Server.CheClusterRoles) > 0 {
+		cheClusterRoles := strings.Split(ctx.CheCluster.Spec.Server.CheClusterRoles, ",")
```
Comment (mmorhun) [inline pkg/deploy/rbac/rbac.go:41]: Truncate spaces?

File: controllers/che/checluster_controller.go (hunk 1 above)
Comment (tolusha) [reply controllers/che/checluster_controller.go:103]: rbac: role-based access control. So it make sense for me to put in that package.

File: pkg/deploy/rbac/gateway_permissions.go (hunk 4 above)
Comment (tolusha) [reply pkg/deploy/rbac/gateway_permissions.go:71]: I think yes

File: pkg/deploy/rbac/rbac.go (hunk 6 above)
Comment (tolusha) [reply pkg/deploy/rbac/rbac.go:41]: It is done below at 43 line

File: pkg/deploy/dev-workspace/dev_workspace_utils.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+func isDevWorkspaceOperatorCSVExists(deployContext *deploy.DeployContext) bool {
+	// If clusterserviceversions resource doesn't exist in cluster DWO as well will not be present
+	if !util.HasK8SResourceObject(deployContext.ClusterAPI.DiscoveryClient, ClusterServiceVersionResourceName) {
+		return false
+	}
+
+	csvList := &operatorsv1alpha1.ClusterServiceVersionList{}
+	err := deployContext.ClusterAPI.Client.List(context.TODO(), csvList, &client.ListOptions{})
```
Comment (AndrienkoAleksandr) [inline pkg/deploy/dev-workspace/dev_workspace_utils.go:45]: Maybe we have to log this an error at least.

File: pkg/deploy/dev-workspace/dev_workspace_utils.go (hunk 3 above)
Comment (tolusha) [reply pkg/deploy/dev-workspace/dev_workspace_utils.go:68]: I've just moved the file. It how it was working before.

File: pkg/deploy/dev-workspace/dev_workspace_utils.go (hunk 7 above)
Comment (tolusha) [reply pkg/deploy/dev-workspace/dev_workspace_utils.go:46]: ok

File: pkg/deploy/dev-workspace/dev_workspace_utils.go (hunk 3 above)
Comment (mmorhun) [reply pkg/deploy/dev-workspace/dev_workspace_utils.go:68]: I see. I think we should reconsider usage of `nonCachingClient` in the main controller, but this is out of scope of this refactoring.

File: pkg/deploy/dev-workspace/dev_workspace_utils.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+func isDevWorkspaceDeploymentExists(deployContext *deploy.DeployContext) (bool, error) {
+	return deploy.Get(deployContext, types.NamespacedName{
+		Namespace: DevWorkspaceNamespace,
+		Name:      DevWorkspaceDeploymentName,
+	}, &appsv1.Deployment{})
+}
+
+func isDevWorkspaceOperatorCSVExists(deployContext *deploy.DeployContext) bool {
+	// If clusterserviceversions resource doesn't exist in cluster DWO as well will not be present
+	if !util.HasK8SResourceObject(deployContext.ClusterAPI.DiscoveryClient, ClusterServiceVersionResourceName) {
```
Comment (AndrienkoAleksandr) [inline pkg/deploy/dev-workspace/dev_workspace_utils.go:41]: Minor, We can simplify:

```
return util.HasK8SResourceObject(deployContext.ClusterAPI.DiscoveryClient, ClusterServiceVersionResourceName)
```

File: pkg/deploy/dev-workspace/dev_workspace_utils.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+			return true
+		}
+	}
+
+	return false
+}
+
+func isWebTerminalSubscriptionExist(deployContext *deploy.DeployContext) (bool, error) {
+	// If subscriptions resource doesn't exist in cluster WTO as well will not be present
+	if !util.HasK8SResourceObject(deployContext.ClusterAPI.DiscoveryClient, SubscriptionResourceName) {
```
Comment (AndrienkoAleksandr) [inline pkg/deploy/dev-workspace/dev_workspace_utils.go:63]: Minor, we can simplify:

```
return !util.HasK8SResourceObject(deployContext.ClusterAPI.DiscoveryClient, SubscriptionResourceName), nil
```

File: pkg/deploy/dev-workspace/dev_workspace.go [hunk 10]
```diff
@@ -55,96 +50,103 @@ var (
...
+			_, err = deploy.Sync(ctx, namespace)
 			if err != nil {
-				return false, err
+				return reconcile.Result{Requeue: true}, false, err
 			}
 		}
 	}
 
 	// check if DW exists on the cluster
 	devWorkspaceWebHookExistedBeforeSync, err := deploy.Get(
```
Comment (AndrienkoAleksandr) [inline pkg/deploy/dev-workspace/dev_workspace.go:140]: I think, here is we ignore error.

**Review by AndrienkoAleksandr — APPROVED** [review]

File: pkg/deploy/dev-workspace/dev_workspace.go [hunk 11]
```diff
@@ -55,96 +50,103 @@ var (
...
-		exists, err := isDevWorkspaceDeploymentExists(deployContext)
+		exists, err := isDevWorkspaceDeploymentExists(ctx)
 		if err != nil {
-			return false, err
+			return reconcile.Result{Requeue: true}, false, err
 		}
 		if !exists {
 			// Don't allow to deploy a new DevWorkspace operator
-			return false, fmt.Errorf("To enable DevWorkspace engine, deploy Eclipse Che from tech-preview channel.")
+			return reconcile.Result{}, false, fmt.Errorf("To enable DevWorkspace engine, deploy Eclipse Che from tech-preview channel.")
```
Comment (AndrienkoAleksandr) [inline pkg/deploy/dev-workspace/dev_workspace.go:96]: Do we need here reconcile.Result{Requeue: true} ?

Comment (tolusha) [reply pkg/deploy/dev-workspace/dev_workspace.go:96]: Requeue won't help so it is not required

**Review by mmorhun — APPROVED** [review]

**Pull request merged**